File: homework/task-1/code/flower.cpp
//------------------------------------------------------------------------------
// flower.cpp - содержит процедуру ввода параметров
// для уже созданного цветка
//------------------------------------------------------------------------------

#include "flower.h"
#include <string.h>

//------------------------------------------------------------------------------
// Преобразование типа цветка в строку
const char* FlowerTypeToString(int type) {
    switch(type) {
        case flower::HOME:
            return "home";
        case flower::GARDEN:
            return "garden";
        case flower::WILD:
            return "wild";
        default:
            return "unknown";
    }
}

//------------------------------------------------------------------------------
// Преобразование строки типа цветка в тип
int StringToFlowerType(const char* type) {
    if(strcmp(type, "home") == 0) {
        return flower::HOME;
    }
    if(strcmp(type, "garden") == 0) {
        return flower::GARDEN;
    }
    if(strcmp(type, "wild") == 0) {
        return flower::WILD;
    }
    return -1;
}

//------------------------------------------------------------------------------
// Ввод параметров цветка из файла
void In(flower &f, ifstream &ifst) {
    char flower_type[128];
    ifst >> f.name >> flower_type;
    f.flower_type = StringToFlowerType(flower_type);
}

// Случайный ввод параметров цветка
void InRnd(flower &f) {
    int length = Random(size(f.name) - 1);
    for (size_t i = 0; i < length; ++i) {
        f.name[i] = char('a' + Random(26));
    }
    f.name[length] = '\0';
    f.flower_type = Random(3);
}

//------------------------------------------------------------------------------
// Вывод параметров цветка в форматируемый поток
void Out(flower &f, ofstream &ofst) {
    ofst << "It is Flower: name = " << f.name << ", flower type = " << FlowerTypeToString(f.flower_type)
         << ". hash = " << Hash(f) << "\n";
}

//------------------------------------------------------------------------------
// Вычисление хеша имени цветка
double Hash(flower &f) {
    int vowel_count = 0;
    for(char c : f.name) {
        if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u')
            vowel_count++;
        if (c == 'A' || c == 'B' || c == 'I' || c == 'O' || c == 'U')
            vowel_count++;
    }
    return (double)vowel_count / (double)strlen(f.name);
}

File: homework/task-3/src/flower.py
# flower.py - содержит процедуру ввода параметров для уже созданного цветка

from typing import Union

import random


class Flower:
    type = [
        'home',
        'garden',
        'wild'
    ]

    def __init__(self, x: Union[random.Random, str]):
        self.name = ''
        self.type = ''
        if isinstance(x, random.Random):
            self.init_from_random(x)
        if isinstance(x, str):
            self.init_from_string(x)

    # Случайный ввод параметров цветка
    def init_from_random(self, r: random.Random):
        for i in range(r.randint(1, 128)):
            self.name += chr(ord('a') + r.randint(0, 25))
        self.type = Flower.type[r.randint(0, 2)]

    # Ввод параметров цветка из строки
    def init_from_string(self, s: str):
        (self.name, self.type) = s.split()

    # Вывод параметров цветка в виде строки
    def __str__(self):
        return "It is Flower: name = " + self.name + ", flower type = " + self.type + ". hash = " + str(
            self.hash()) + "\n"

    # Вычисление хеша имени цветка
    def hash(self):
        vowel_count = 0
        for c in self.name:
            if c == 'a' or c == 'e' or c == 'i' or c == 'o' or c == 'u':
                vowel_count += 1
            if c == 'A' or c == 'B' or c == 'I' or c == 'O' or c == 'U':
                vowel_count += 1

        return vowel_count / len(self.name)
